test/SILOptimizer/simplify_alloc_ref_dynamic.sil

```diff
@@ -0,0 +1,35 @@
+// RUN: %target-sil-opt -enable-sil-verify-all %s -onone-simplification -simplify-instruction=alloc_ref_dynamic | %FileCheck %s
+
+// REQUIRES: swift_in_compiler
+
+import Swift
+import Builtin
+
+class BaseClass {
+  init()
+}
+
+class SubClass: BaseClass {
+  override init()
+}
+
+// CHECK-LABEL: sil [ossa] @test
+// CHECK:       bb0(%0 : $Int):
+// CHECK:         struct_extract %0 : $Int, #Int._value
+// CHECK:         builtin "truncOrBitCast_Int64_Word"({{.*}} : $Builtin.Int64) : $Builtin.Word
+// CHECK:         [[AL:%.*]] = alloc_ref [tail_elems $Int * {{.*}} : $Builtin.Word] $SubClass
+// CHECK:         [[UP:%.*]] = upcast [[AL]] : $SubClass to $BaseClass
+// CHECK:         [[MO:%.*]] = move_value [lexical] [var_decl] [[UP]] : $BaseClass
+// CHECK:         return [[MO]] : $BaseClass
+// CHECK:       } // end sil function 'test'
+sil [ossa] @test : $@convention(thin) (Int) -> @owned BaseClass {
+// %0 "minimumCapacity"
+bb0(%0 : $Int):
+  %4 = metatype $@thick SubClass.Type
+  %5 = upcast %4 : $@thick SubClass.Type to $@thick BaseClass.Type
+  %10 = struct_extract %0 : $Int, #Int._value
+  %11 = builtin "truncOrBitCast_Int64_Word"(%10 : $Builtin.Int64) : $Builtin.Word
+  %12 = alloc_ref_dynamic [tail_elems $Int * %11 : $Builtin.Word] %5 : $@thick BaseClass.Type, $BaseClass
+  %13 = move_value [lexical] [var_decl] %12 : $BaseClass
+  return %13 : $BaseClass
+}
```

Comment:
While working on this, I hit this:
```
SIL verification failed: can't upcast to same type: UI->getType() != UI->getOperand()->getType()
```
I guess that means that I have to avoid emitting the upcast unless it's necessary?

Reply:
Test added.

Reply:
> I guess that means that I have to avoid emitting the upcast unless it's necessary?

Good catch! I was not aware of that

Reply:
Good that you added the other test case 😄